### DISCUSSION TITLE
v3: Exports can use git for distribution

### DISCUSSION THREAD

**Discussion by msimerson** [NicTool/NicTool#285] [Ideas]
For most NS types, the default mechanism NicTool uses to distribute NS data is rsync. The export process is summarized as:

1. export zones to files
2. use rsync to distribute files to the [usually remote] nameserver(s)
3. restart/reload remote NS

Rsync is used because _usually_ it's very bandwidth efficient. However, when the number of zones is very large, there are long delays as rsync must stat every zone file on each end, share the catalog of files, and then do the updates. Git may be more efficient. The export process might look like:

1. export zones to local files, in a git repo
2. `git commit -m 'updates' && git push`
3. ssh to the remote NS and run `git pull && service [NStype] {restart | reload | reload $zone}`

The setup would be similar (ssh key based auth). The principal difference is that each remote NS would need access _to_ the git repo where the DNS data is stored. This could be via http or SSH.

**Comment by rduffner** [comment]
rsync is much simpler and less prone to errors.

The upside of git pull is that you don't need an ssh-trust-relationship between the nictool-client and the remote DNS.

**Reply by msimerson** [reply]
> rsync is much simpler

Agreed. Simple is good. Another approach to exports might be a chained set of CLI commands (similar to CI systems like GitHub Actions) where you declare the steps and they are executed. That also lets one integrate features like DNSSEC signing before publishing.

**Reply by meersau** [reply]
Every setup is unique. We have Nictool in our internal security zone and push the exported files to our name server. 

It would be great if the export is not bind to any specific technology. A CLI command like nt_export.pl is a great tool to use.
